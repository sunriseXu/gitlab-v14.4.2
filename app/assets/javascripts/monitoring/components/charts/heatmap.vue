<script>
import { GlHeatmap } from '@gitlab/ui/dist/charts';
import { formatDate, timezones, formats } from '../../format_date';
import { graphDataValidatorForValues } from '../../utils';

export default {
  components: {
    GlHeatmap,
  },
  props: {
    graphData: {
      type: Object,
      required: true,
      validator: graphDataValidatorForValues.bind(null, false),
    },
    timezone: {
      type: String,
      required: false,
      default: timezones.LOCAL,
    },
  },
  data() {
    return {
      width: 0,
    };
  },
  computed: {
    chartData() {
      return this.metrics.result.reduce(
        (acc, result, i) => [...acc, ...result.values.map((value, j) => [i, j, value[1]])],
        [],
      );
    },
    xAxisName() {
      return this.graphData.xLabel || '';
    },
    yAxisName() {
      return this.graphData.y_label || '';
    },
    xAxisLabels() {
      return this.metrics.result.map((res) => Object.values(res.metric)[0]);
    },
    yAxisLabels() {
      return this.result.values.map((val) => {
        const [yLabel] = val;

        return formatDate(new Date(yLabel), {
          format: formats.shortTime,
          timezone: this.timezone,
        });
      });
    },
    result() {
      return this.metrics.result[0];
    },
    metrics() {
      return this.graphData.metrics[0];
    },
  },
};
</script>
<template>
  <gl-heatmap
    ref="heatmapChart"
    v-bind="$attrs"
    :responsive="true"
    :data-series="chartData"
    :x-axis-name="xAxisName"
    :y-axis-name="yAxisName"
    :x-axis-labels="xAxisLabels"
    :y-axis-labels="yAxisLabels"
    :width="width"
  />
</template>
