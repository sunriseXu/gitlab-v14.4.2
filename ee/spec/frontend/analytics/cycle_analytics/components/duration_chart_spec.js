import { GlIcon } from '@gitlab/ui';
import { GlLineChart } from '@gitlab/ui/dist/charts';
import { shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import {
  DURATION_STAGE_TIME_DESCRIPTION,
  DURATION_TOTAL_TIME_DESCRIPTION,
  DURATION_STAGE_TIME_NO_DATA,
  DURATION_TOTAL_TIME_NO_DATA,
} from 'ee/analytics/cycle_analytics/constants';
import DurationChart from 'ee/analytics/cycle_analytics/components/duration_chart.vue';
import ChartSkeletonLoader from '~/vue_shared/components/resizable_chart/skeleton_loader.vue';
import {
  allowedStages as stages,
  durationChartPlottableData as durationData,
  durationDataSeries,
  durationDataNullSeries,
} from '../mock_data';

Vue.use(Vuex);

const fakeStore = ({ initialGetters, initialState, rootGetters, rootState }) =>
  new Vuex.Store({
    state: {
      ...rootState,
    },
    getters: {
      isOverviewStageSelected: () => true,
      ...rootGetters,
    },
    modules: {
      durationChart: {
        namespaced: true,
        getters: {
          durationChartPlottableData: () => durationData,
          ...initialGetters,
        },
        state: {
          isLoading: false,
          ...initialState,
        },
      },
    },
  });

function createComponent({
  stubs = {},
  initialState = {},
  initialGetters = {},
  rootGetters = {},
  rootState = {},
  props = {},
} = {}) {
  return shallowMount(DurationChart, {
    store: fakeStore({ initialState, initialGetters, rootGetters, rootState }),
    propsData: {
      stages,
      ...props,
    },
    stubs: {
      ChartSkeletonLoader: true,
      ...stubs,
    },
  });
}

describe('DurationChart', () => {
  let wrapper;

  const findContainer = (_wrapper) => _wrapper.find('[data-testid="vsa-duration-chart"]');
  const findChartDescription = (_wrapper) => _wrapper.findComponent(GlIcon);
  const findDurationChart = (_wrapper) => _wrapper.findComponent(GlLineChart);
  const findLoader = (_wrapper) => _wrapper.findComponent(ChartSkeletonLoader);

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe('with the overiew stage selected', () => {
    beforeEach(() => {
      wrapper = createComponent({});
    });

    it('renders the duration chart', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('correctly sets the chart options data property', () => {
      const chartDataProps = findDurationChart(wrapper).props('data');
      const [dataSeries, nullSeries] = chartDataProps;
      const expectedParams = ['areaStyle', 'data', 'itemStyle', 'lineStyle', 'name', 'showSymbol'];

      expectedParams.forEach((param) => {
        expect(dataSeries[param]).toEqual(durationDataSeries[param]);
        expect(nullSeries[param]).toEqual(durationDataNullSeries[param]);
      });
    });

    it('renders the chart', () => {
      expect(findDurationChart(wrapper).exists()).toBe(true);
    });

    it('renders the chart description', () => {
      expect(findChartDescription(wrapper).attributes('title')).toBe(
        DURATION_TOTAL_TIME_DESCRIPTION,
      );
    });

    describe('with no chart data', () => {
      beforeEach(() => {
        wrapper = createComponent({
          initialGetters: {
            durationChartPlottableData: () => [],
          },
        });
      });

      it('renders the no data available message', () => {
        expect(findContainer(wrapper).text()).toContain(DURATION_TOTAL_TIME_NO_DATA);
      });
    });
  });

  describe('with a value stream stage selected', () => {
    const [selectedStage] = stages;

    beforeEach(() => {
      wrapper = createComponent({
        rootState: {
          selectedStage,
        },
        rootGetters: {
          isOverviewStageSelected: () => false,
        },
      });
    });

    it('renders the chart', () => {
      expect(findDurationChart(wrapper).exists()).toBe(true);
    });

    it('renders the stage title', () => {
      expect(wrapper.text()).toContain(`Stage time: ${selectedStage.title}`);
    });

    it('sets the chart data', () => {
      expect(findDurationChart(wrapper).props('data')).toEqual([
        expect.objectContaining(durationDataSeries),
        durationDataNullSeries,
      ]);
    });

    it('renders the chart description', () => {
      expect(findChartDescription(wrapper).attributes('title')).toBe(
        DURATION_STAGE_TIME_DESCRIPTION,
      );
    });

    describe('with no chart data', () => {
      beforeEach(() => {
        wrapper = createComponent({
          initialGetters: {
            durationChartPlottableData: () => [[new Date(), null]],
          },
          rootState: {
            selectedStage,
          },
          rootGetters: {
            isOverviewStageSelected: () => false,
          },
        });
      });

      it('renders the no data available message', () => {
        expect(findContainer(wrapper).text()).toContain(DURATION_STAGE_TIME_NO_DATA);
      });
    });
  });

  describe('when isLoading=true', () => {
    beforeEach(() => {
      wrapper = createComponent({ initialState: { isLoading: true } });
    });

    it('renders a loader', () => {
      expect(findLoader(wrapper).exists()).toBe(true);
    });
  });
});
