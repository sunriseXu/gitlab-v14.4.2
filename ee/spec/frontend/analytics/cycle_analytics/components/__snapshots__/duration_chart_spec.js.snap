// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationChart with the overiew stage selected renders the duration chart 1`] = `
<div
  class="gl-display-flex gl-flex-direction-column"
  data-testid="vsa-duration-chart"
>
  <h4
    class="gl-mt-0"
  >
    
    Total timeÂ 
    <gl-icon-stub
      name="information-o"
      size="16"
      title="The total time items spent across each value stream stage. Data limited to items completed within this date range."
    />
  </h4>
   
  <gl-line-chart-stub
    annotations=""
    data="[object Object],[object Object]"
    formattooltiptext="[Function]"
    legendaveragetext="Avg"
    legendcurrenttext="Current"
    legendlayout="inline"
    legendmaxtext="Max"
    legendmintext="Min"
    option="[object Object]"
    thresholds=""
  />
</div>
`;
